File: stages-backend/src/main/java/com/gestion/stages_backend/offres/application/service/OffreService.java
package com.gestion.stages_backend.offres.application.service;

import com.gestion.stages_backend.identity.domain.model.User;
import com.gestion.stages_backend.identity.infrastructure.persistence.JpaUserRepository;
import com.gestion.stages_backend.offres.application.dto.OffreRequest;
import com.gestion.stages_backend.offres.application.dto.OffreResponse;
import com.gestion.stages_backend.offres.application.mapper.OffreMapper;
import com.gestion.stages_backend.offres.domain.model.Domaine;
import com.gestion.stages_backend.offres.domain.model.Offre;
import com.gestion.stages_backend.offres.domain.model.StatutOffre;
import com.gestion.stages_backend.offres.infrastructure.persistence.JpaOffreRepository;
import com.gestion.stages_backend.shared.application.dto.PageResponse;
import com.gestion.stages_backend.shared.infrastructure.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class OffreService {

    private final JpaOffreRepository offreRepository;
    private final JpaUserRepository userRepository;
    private final OffreMapper offreMapper;

    @Transactional
    public OffreResponse publier(OffreRequest request) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        User entreprise = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("Entreprise introuvable"));
        Offre offre = offreMapper.toEntity(request);
        offre.setEntreprise(entreprise);
        return offreMapper.toResponse(offreRepository.save(offre));
    }

    @Transactional(readOnly = true)
    public PageResponse<OffreResponse> rechercher(Domaine domaine, String localisation,
                                                   Integer dureeMois, Pageable pageable) {
        return PageResponse.of(
                offreRepository.findByStatut(StatutOffre.ACTIVE, pageable)
                        .map(offreMapper::toResponse));
    }

    @Transactional(readOnly = true)
    public OffreResponse findById(Long id) {
        return offreMapper.toResponse(offreRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Offre", id)));
    }

    @Transactional(readOnly = true)
    public PageResponse<OffreResponse> mesOffres(Pageable pageable) {
        String email = SecurityContextHolder.getContext().getAuthentication().getName();
        User entreprise = userRepository.findByEmail(email)
                .orElseThrow(() -> new ResourceNotFoundException("Entreprise introuvable"));
        return PageResponse.of(
                offreRepository.findByEntrepriseId(entreprise.getId(), pageable)
                        .map(offreMapper::toResponse));
    }

    @Transactional
    public OffreResponse archiver(Long id) {
        Offre offre = offreRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Offre", id));
        offre.setStatut(StatutOffre.ARCHIVEE);
        return offreMapper.toResponse(offreRepository.save(offre));
    }
}

File: stages-frontend/src/app/core/services/offre.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { ApiResponse, PageResponse } from "../models/api.model";
import { OffreRequest, OffreResponse, Domaine } from "../models/offre.model";

@Injectable({ providedIn: "root" })
export class OffreService {
  private api = "http://localhost:8082/api/offres";
  constructor(private http: HttpClient) {}

  rechercher(domaine?: Domaine, localisation?: string, dureeMois?: number, page = 0, size = 10) {
    let params = new HttpParams().set("page", page).set("size", size);
    if (domaine)      params = params.set("domaine", domaine);
    if (localisation) params = params.set("localisation", localisation);
    if (dureeMois)    params = params.set("dureeMois", dureeMois);
    return this.http.get<ApiResponse<PageResponse<OffreResponse>>>(this.api, { params });
  }

  findById(id: number) {
    return this.http.get<ApiResponse<OffreResponse>>(`${this.api}/${id}`);
  }

  publier(req: OffreRequest) {
    return this.http.post<ApiResponse<OffreResponse>>(this.api, req);
  }

  mesOffres(page = 0, size = 50) {
    const params = new HttpParams().set("page", page).set("size", size);
    return this.http.get<ApiResponse<PageResponse<OffreResponse>>>(`${this.api}/mes-offres`, { params });
  }

  archiver(id: number) {
    return this.http.patch<ApiResponse<OffreResponse>>(`${this.api}/${id}/archiver`, {});
  }
}
